packages/sync/src/range.rs

```diff
@@ -0,0 +1,399 @@
+
+#[derive(Debug, Error, PartialEq, Eq)]
+pub enum RangeError {
+    #[error("Underflow minimum value")]
+    Underflow,
+    #[error("Overflow maximum value")]
+    Overflow,
+}
+
+impl<T> ValueRange<T>
+where
+    T: Copy,
+{
+    pub fn new(value: T) -> Self {
+        Self(value, value)
+    }
+
+    pub fn min(&self) -> T {
+        self.0
+    }
+
+    pub fn max(&self) -> T {
+        self.1
+    }
+}
+
+/// Problem: We have a list of ValueRanges, and we want to know the maximum value.
+/// This is not one clear value, as we consider the maximum if all commit and maximum if all rollback.
+/// The result is the range of possible maximum values (different than spread)  
+pub fn max_range<'a, I, T>(iter: I) -> ValueRange<T>
+where
+    I: Iterator<Item = &'a ValueRange<T>> + 'a,
+    T: Ord + Copy + Default + 'a,
+{
+    iter.copied()
+        .reduce(|acc, x| {
+            ValueRange(
+                std::cmp::max(acc.min(), x.min()),
+                std::cmp::max(acc.max(), x.max()),
+            )
+        })
+        .unwrap_or_default()
+}
+
+/// Problem: We have a list of ValueRanges, and we want to know the minimum value.
+/// This is not one clear value, as we consider the minimum if all commit and minimum if all rollback.
+/// The result is the range of possible minimum values (different than spread)  
+pub fn min_range<'a, I, T>(iter: I) -> ValueRange<T>
+where
+    I: Iterator<Item = &'a ValueRange<T>> + 'a,
+    T: Ord + Copy + Default + 'a,
+{
+    iter.copied()
+        .reduce(|acc, x| {
+            ValueRange(
+                std::cmp::min(acc.min(), x.min()),
+                std::cmp::min(acc.max(), x.max()),
+            )
+        })
+        .unwrap_or_default()
+}
+
+/// Captures the spread from the lowest low to the highest high
+pub fn spread<'a, I, T>(iter: I) -> ValueRange<T>
+where
+    I: Iterator<Item = &'a ValueRange<T>> + 'a,
+    T: Ord + Copy + Default + 'a,
+{
+    iter.copied()
+        .reduce(|acc, x| {
+            ValueRange(
+                std::cmp::min(acc.min(), x.min()),
+                std::cmp::max(acc.max(), x.max()),
+            )
+        })
+        .unwrap_or_default()
+}
+
+impl<T, U> Mul<U> for ValueRange<T>
+where
+    T: Mul<U, Output = T>,
+    U: Copy,
+{
+    type Output = ValueRange<T>;
+
+    fn mul(self, rhs: U) -> Self::Output {
+        ValueRange(self.0 * rhs, self.1 * rhs)
+    }
+}
+
+impl<T, U> Mul<U> for &ValueRange<T>
+where
+    T: Mul<U, Output = T> + Copy,
+    U: Copy,
+{
+    type Output = ValueRange<T>;
+
+    fn mul(self, rhs: U) -> Self::Output {
+        ValueRange(self.0 * rhs, self.1 * rhs)
+    }
+}
+
+impl<T> ValueRange<T>
+where
+    T: Add<Output = T> + Sub<Output = T> + Ord + Copy,
+{
+    /// This is a check for calling code if it wishes to change the (externally defined) maximum for the range.
+    /// Usage is eg modifying collateral while the range is total liens on the collateral
+    pub fn valid_max(&self, new_max: T) -> bool {
```

Comment:
This looks good, but the API is still confusing.

I would suggest using `is`, for queries that don't modify state:
```suggestion
    pub fn is_valid_max(&self, new_max: T) -> bool {
```

Reply:
Good idea